::::: {.content-visible when-meta="inspect"}

## Data inspection

It's important to ensure we have enough sequencing reads with sufficient information
to accurately assemble the species' genome.

### Sequence data quantity

* Do we have enough total sequence data?
* Do we have enough long reads?

#### HiFi

```{python}
#| output: asis
if inspect and glob.glob(f"{log_path}/*_hifi.tsv"):
    displayTsvAsTable(f"{log_path}/*_hifi.tsv")
```

::: {.callout-note title="Expectation" collapse="true"}
For long read data (PacBio, ONT) we expect to see 15-20x coverage per haplotype, for a good assembly.
:::

#### Hi-C

```{python}
#| output: asis
if inspect and glob.glob(f"{log_path}/*_hic.tsv"):
    displayTsvAsTable(f"{log_path}/*_hic.tsv")
```

```{python}
if inspect and "fastqc" in multiqc.list_modules():
    display(multiqc.get_plot("fastqc", "Sequence Counts").show())
else:
    print("No Hi-C data available")
```

::: {.callout-note title="Expectation" collapse="true"}
Arima-HiC (Arima Genomics)
: Duplication levels expected to be ~10-30%.

Dovetail Omni-C (Cantata Bio)
: Duplication levels expected to be ~30-80%.
:::

```{python}
if inspect and "fastqc" in multiqc.list_modules():
    display(multiqc.get_plot("fastqc", "Sequence Duplication Levels").show())
```

#### HiFi vs Hi-C

* Do our libraries have similar data content?
```{python}
if inspect and glob.glob(f"{log_path}/*_katcomp.st.png"):
    display(Image(filename=glob.glob(f"{log_path}/*_katcomp.st.png")[0]))
else:
    print("A comparison of HiFi vs Hi-C data is not available")
```

### Genomic characteristics

K-mer analysis provides valuable insights into the genomic characteristics
of a sequenced organism. Tools such as GenomeScope2 / GeneScopeFK, and
Smudgeplot can be used to infer the organism's ploidy and coverage by
analyzing its k-mer content.

```{python}
#| output: asis
if inspect and glob.glob(f"{log_path}/*_log_plot.png"):
    displayGenomeScopePlots()
```

::: {.callout-note title="Interpretation" collapse="true"}
A Kmer histogram counts the frequency of k-mers in the data. Some parts of the genome are unique, and
there are expected to be *coverage* times copies of these k-mers in the data. If the organism is diploid,
the regions unique to a haplotype should have *coverage* times copies of these k-mers, and regions shared
between both haplotypes should have *coverage* times two copies of these k-mers. This leads to formations
of peaks in the k-mer histogram. A haploid organism should have one peak, while a diploid should have two peaks.
:::

#### SmudgePlot

Smudgeplot is used to infer the ploidy of the organism.

```{python}
if inspect and glob.glob(f"{log_path}/*_ploidyplot.st.png"):
    display(Image(filename=glob.glob(f"{log_path}/*_ploidyplot.st.png")[0]))
else:
    print("Ploidyplot not available")
```

#### GC content

```{python}
if inspect and glob.glob(f"{log_path}/*_katgc.st.png"):
    display(Image(filename=glob.glob(f"{log_path}/*_katgc.st.png")[0]))
else:
    print("KAT GC not available")
```

::: {.callout-note title="Expectation" collapse="true"}
* Unexpected GC content value can indicate contamination.
* Unexpected blobs(?) in the plot can indicate contamination.
:::

#### Hi-C

```{python}
if inspect and "fastqc" in multiqc.list_modules():
    display(multiqc.get_plot("fastqc", "Per Sequence GC Content").show())
```

::: {.callout-note title="Expectation" collapse="true"}
The peak GC content on the x-axis should be the same value
as the peak on the HiFi KatGC plot on the y-axis.
:::

### Synthetic sequence content

Do we see indications of synthetic sequences such as adaptors?

#### HiFi

TODO: FCS

#### Hi-C
```{python}
if inspect and "fastqc" in multiqc.list_modules():
    display(multiqc.get_plot("fastqc", "Adapter Content").show())
```

:::::
