Changes in module 'nf-core/pairtools/split'
--- modules/nf-core/pairtools/split/main.nf
+++ modules/nf-core/pairtools/split/main.nf
@@ -11,10 +11,15 @@
 
     input:
     tuple val(meta), path(pairs)
+    tuple val(meta), path(fasta)
+    val   sort_bam 
 
     output:
     tuple val(meta), path("*.split.pairs.gz"), emit:pairs
-    tuple val(meta), path("*.bam")           , emit:bam   , optional:true
+    tuple val(meta), path("*.bam")           , emit:bam  , optional:true
+    tuple val(meta), path("*.cram")          , emit: cram, optional:true
+    tuple val(meta), path("*.crai")          , emit: crai, optional:true
+    tuple val(meta), path("*.csi")           , emit: csi , optional:true
     path("versions.yml")                     , emit:versions
 
     when:
@@ -22,18 +27,30 @@
 
     script:
     def args = task.ext.args ?: ''
+    def args2 = task.ext.args2 ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def samtools_command = sort_bam ? 'sort' : 'view'
+
+    def extension_pattern = /(--output-fmt|-O)+\s+(\S+)/
+    def extension_matcher =  (args2 =~ extension_pattern)
+    def extension = extension_matcher.getCount() > 0 ? extension_matcher[0][2].toLowerCase() : "bam"
+    def reference = fasta && extension=="cram"  ? "--reference ${fasta}" : ""
+    if (!fasta && extension=="cram") error "Fasta reference is required for CRAM output"
+
     """
     export MPLCONFIGDIR=tmp
     pairtools split \
         --nproc-in ${task.cpus} --nproc-out ${task.cpus} \
         --output-pairs ${prefix}.split.pairs.gz \
+        --output-sam - \
         ${args} \
-        ${pairs}
+        ${pairs} \
+        | samtools $samtools_command $args2 -@ $task.cpus ${reference} -o ${prefix}.split.pairs.${extension} -
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         pairtools: \$(pairtools --version 2>&1 | sed 's/pairtools, version //')
+        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
     END_VERSIONS
     """
     stub:
@@ -45,6 +62,7 @@
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         pairtools: \$(pairtools --version | sed 's/pairtools, version //')
+        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
     END_VERSIONS
     """
 

************************************************************
