Changes in module 'nf-core/pairtools/sort'
--- modules/nf-core/pairtools/sort/environment.yml
+++ modules/nf-core/pairtools/sort/environment.yml
@@ -2,5 +2,5 @@
   - conda-forge
   - bioconda
 dependencies:
-  - bioconda::pairtools=1.0.2
+  - bioconda::pairtools=1.1.0
   - conda-forge::numpy=1.23

--- modules/nf-core/pairtools/sort/main.nf
+++ modules/nf-core/pairtools/sort/main.nf
@@ -6,8 +6,8 @@
     // Not an issue with the biocontainers because they were built prior to numpy 1.24
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/pairtools:1.0.2--py39h2a9f597_0' :
-        'biocontainers/pairtools:1.0.2--py39h2a9f597_0' }"
+        'https://depot.galaxyproject.org/singularity/pairtools:1.1.0--py39hd5a99d8_2' :
+        'biocontainers/pairtools:1.1.0--py39hd5a99d8_2' }"
 
     input:
     tuple val(meta), path(input)

************************************************************
