Changes in module 'nf-core/star/genomegenerate'
--- modules/nf-core/star/genomegenerate/main.nf
+++ modules/nf-core/star/genomegenerate/main.nf
@@ -29,7 +29,7 @@
             --runMode genomeGenerate \\
             --genomeDir star/ \\
             --genomeFastaFiles $fasta \\
-            --sjdbGTFfile $gtf \\
+            ${ gtf ? "--sjdbGTFfile $gtf": '' } \\
             --runThreadN $task.cpus \\
             $memory \\
             $args
@@ -51,7 +51,7 @@
             --runMode genomeGenerate \\
             --genomeDir star/ \\
             --genomeFastaFiles $fasta \\
-            --sjdbGTFfile $gtf \\
+            ${ gtf ? "--sjdbGTFfile $gtf" : '' } \\
             --runThreadN $task.cpus \\
             --genomeSAindexNbases \$NUM_BASES \\
             $memory \\

************************************************************
