Changes in module 'nf-core/deepvariant'
--- modules/nf-core/deepvariant/main.nf
+++ modules/nf-core/deepvariant/main.nf
@@ -2,7 +2,7 @@
     tag "$meta.id"
     label 'process_high'
 
-    container "nf-core/deepvariant:1.5.0"
+    container = "docker.io/google/deepvariant:CL602468145"
 
     input:
     tuple val(meta), path(input), path(index), path(intervals)
@@ -27,7 +27,7 @@
     }
     def args = task.ext.args ?: ''
     prefix = task.ext.prefix ?: "${meta.id}"
-    def regions = intervals ? "--regions=${intervals}" : ""
+    def regions = intervals ? "--regions=./${intervals}" : ""
 
     """
     /opt/deepvariant/bin/run_deepvariant \\
@@ -37,7 +37,7 @@
         --output_gvcf=${prefix}.g.vcf.gz \\
         ${args} \\
         ${regions} \\
-        --intermediate_results_dir=. \\
+        --intermediate_results_dir=intermediate_results \\
         --num_shards=${task.cpus}
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
