Changes in module 'nf-core/star/align'
--- modules/nf-core/star/align/main.nf
+++ modules/nf-core/star/align/main.nf
@@ -9,8 +9,8 @@
 
     input:
     tuple val(meta), path(reads, stageAs: "input*/*")
-    tuple val(meta2), path(index)
-    tuple val(meta3), path(gtf)
+    path(index)
+    path(gtf)
     val star_ignore_sjdbgtf
     val seq_platform
     val seq_center

************************************************************
